internal/tracer/reader.go

```diff
@@ -116,6 +118,9 @@ type dataTracer struct {
 }
 
 func (d *dataTracer) trace(data []byte) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
```

Comment:
I happened to trip the race detector when running tests at one point, perhaps because of the add'l test cases added to the `TestRun` test in `connectconformance_test.go`.

I forgot that the body of an `http.Request` needs to be thread-safe to at least support a call to `Close` being concurrent with a call to `Read`. And that was not the case before, but now it is safe, due to this locking.